pkg/csi/cinder/controllerserver.go

```diff
@@ -335,73 +381,208 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
+	if snapshotType != "snapshot" && snapshotType != "backup" {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot type must be 'backup', 'snapshot' or not defined")
+	}
+
+	// Prechecks in case of a backup
+	if snapshotType == "backup" {
+
+		// Get a list of backups with the provided name
+		backups, err = cs.Cloud.ListBackups(filters)
+		if err != nil {
+			klog.Errorf("Failed to query for existing Backup during CreateSnapshot: %v", err)
+			return nil, status.Error(codes.Internal, "Failed to get backups")
+		}
+		// If more than one backup with the provided name exists, fail
+		if len(backups) > 1 {
+			klog.Errorf("found multiple existing backups with selected name (%s) during create", name)
+			return nil, status.Error(codes.Internal, "Multiple backups reported by Cinder with same name")
+		}
+
+		if len(backups) == 1 {
+			backup = &backups[0]
+
+			// Verify the existing backup has the same VolumeID, otherwise it belongs to another volume
+			if backup.VolumeID != volumeID {
+				return nil, status.Error(codes.AlreadyExists, "Backup with given name already exists, with different source volume ID")
+			}
+
+			// If a backup of the volume already exists, skip creating the snapshot
+			backupAlreadyExists = true
+			klog.V(3).Infof("Found existing backup %s from volume with ID: %s", name, volumeID)
+		}
+	}
+
+	// Create the snapshot if the backup does not already exist and wait for it to be ready
+	if !backupAlreadyExists {
+		snap, err = cs.createSnapshot(name, volumeID, req.Parameters)
+		if err != nil {
+			return nil, err
+		}
+
+		ctime = timestamppb.New(snap.CreatedAt)
+		if err = ctime.CheckValid(); err != nil {
+			klog.Errorf("Error to convert time to timestamp: %v", err)
+		}
+
+		err = cs.Cloud.WaitSnapshotReady(snap.ID)
+		if err != nil {
+			klog.Errorf("Failed to WaitSnapshotReady: %v", err)
+			return nil, status.Errorf(codes.Internal, "CreateSnapshot failed with error %v", err)
+		}
+
+		snapSize = snap.Size
+	}
+
+	if snapshotType == "snapshot" {
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SnapshotId:     snap.ID,
+				SizeBytes:      int64(snap.Size * 1024 * 1024 * 1024),
+				SourceVolumeId: snap.VolumeID,
+				CreationTime:   ctime,
+				ReadyToUse:     true,
+			},
+		}, nil
+	}
+
+	// If snapshotType is 'backup', create a backup from the snapshot and delete the snapshot.
+	if snapshotType == "backup" {
+
+		if !backupAlreadyExists {
+			backup, err = cs.createBackup(name, volumeID, snap, req.Parameters)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		ctime = timestamppb.New(backup.CreatedAt)
+		if err := ctime.CheckValid(); err != nil {
+			klog.Errorf("Error to convert time to timestamp: %v", err)
+		}
+
+		// Get the max duration to wait in seconds per GB of snapshot
+		backupMaxDurationSecondsPerGB, err = strconv.Atoi(openstack.BackupMaxDurationPerGB)
+		if err != nil {
+			klog.V(5).Infof("Set backup-max-duration-seconds-per-gb to 20 due to: %v", err)
```

Comment:
```suggestion
			klog.V(5).Infof("Set backup-max-duration-seconds-per-gb to %d due to parsing error: %v", openstack.BackupMaxDurationSecondsPerGBDefault, err)
```

Reply:
Actually I think that it makes sense to parse the `BackupMaxDurationPerGB` during the controller startup and fail with an error, if the value cannot be parsed.

Reply:
here I was following what they do for the force create flag for snapshots:
```
func (os *OpenStack) CreateSnapshot(name, volID string, tags map[string]string) (*snapshots.Snapshot, error) {

	force := false
	// if no flag given, then force will be false by default
	// if flag it given , check it
	if item, ok := (tags)[SnapshotForceCreate]; ok {
		var err error
		force, err = strconv.ParseBool(item)
		if err != nil {
			klog.V(5).Infof("Make force create flag to false due to: %v", err)
		}

		delete(tags, SnapshotForceCreate)
	}
...
```
Where if the parsing of an argument that comes from the `volumeSnapshotClass` fails, it goes to the default value instead of failing.

However now that I think about it, in the cases where you meant to set force create flag to true and it did not parse correctly, the snapshot would have failed anyway because it is necessary to have it set to true to create the snapshot...

I can make it fail at the start of `CreateSnapshot()`, but not at controller startup since you only get this value from `req.parameters`. It's from the `volumeSnapshotClass`.